### ISSUE TITLE
Type annotations for query parameters

### ISSUE THREAD

**Issue by kyleconroy** [sqlc-dev/sqlc#2800]
I propose adding type annotations for query parameters. Users will no longer need to cast parameters to the desired type. This proposal builds on Andrew's query annotation
work for `vet`.

## Unified syntax

We added the `@sqlc-vet-disable` annotation to disable vet rules on a per-query
basis. We can extend this syntax to support other per-query configuration
options.

The current syntax for query name and command are different, so we'll
standardize on the `@` prefix. This will be the new, preferred syntax for name
and command.

```sql
-- name: GetAuthor :one
SELECT * FROM authors
WHERE id = $1 LIMIT 1;

-- becomes

-- @name GetAuthor :one
SELECT * FROM authors
WHERE id = $1 LIMIT 1;
```

The existing syntax will continue to work, but it will be an error to use both
annotations on a single query.

```sql
-- INVALID!
-- name: GetAuthor :one
-- @name GetAuthor :one
SELECT * FROM authors
WHERE id = $1 LIMIT 1;
```

## Query command

Today, queries must have a name and a command. With the new syntax, the
command option will default to `exec`.

```sql
-- These two annotations are the same

-- @name DeleteAuthor :exec
DELETE FROM authors
WHERE id = $1;

-- @name DeleteAuthor
DELETE FROM authors
WHERE id = $1;
```

### Validation

sqlc will delegate command validation to codegen plugins, allowing plugins to implement new commands without having
to merge anything into sqlc.

If you want to still validate those, you can simulate the current behavior by
using this vet rule.

```yaml
rules:
  - name: validate-command
    rule: |
      !(query.cmd in ["exec", "one", "many", "execrows", "execlastid", "execresult", "batchone", "batchmany", "batchexec"])
```

### Type annotations for query parameters

The `@param` annotation supports passing type information without having to use
a cast. 

```sql
-- @param name type
```

Casts are required today when sqlc infers an incorrect type, but these
casts are passed down to the engine itself, possibly hurting performance. 
For example, this cast is required to get sqlc working correctly, but isn't needed at runtime.

```sql
-- @name ListAuthorsByIDs :many
SELECT * FROM authors
WHERE id = ANY(@ids::bigint[]);
```
Here's what it looks like with the new syntax. The type annotation is
engine-specific and is the same that you'd pass to CAST or CREATE TABLE.

```sql
-- @name ListAuthorsByIDs :many
-- @param ids BIGINT[] NOT NULL
SELECT * FROM authors
WHERE id = ANY(@ids);
```

If a paramater has a type annotation, that will be used instead of inferring the
type from the query.

### NULL values

Types in annotations are NULL by default, matching behavior in column definitions and SQL function definitions. If you'd like your parameter to be not null, add the `NOT NULL` syntax. 

```sql
-- @name CreateAuthor :one
-- @param name TEXT NOT NULL
-- @param bio TEXT
INSERT INTO authors (
  name, bio
) VALUES (
  @name, @bio
)
RETURNING *;
```

### Positional parameters

If your parameters do not have a given name, use the `$n` or `?n` syntax to add a type annotation.

```sql
-- @name CreateAuthor :one
-- @param $1 TEXT NULL
-- @param $2 TEXT NULL
INSERT INTO authors (
  name, bio
) VALUES (
  $1, $2
)
RETURNING *;
```

```sql
-- @name CreateAuthor :one
-- @param ?1 TEXT NULL
-- @param ?2 TEXT NULL
INSERT INTO authors (
  name, bio
) VALUES (
  ?1, ?2
)
RETURNING *;
```

## Why comments?

We're using comments instead of `sqlc.*` functions to avoid engine-specific parsing issues. For example, we've run into issues with the MySQL parser not support functions in certain parameter locations.

## Full example

This is the normal example from the playground using the new syntax.

```sql
-- @name GetAuthor :one
-- @param id BIGINT NOT NULL
SELECT * FROM authors
WHERE id = @id LIMIT 1;

-- @name ListAuthors :many
SELECT * FROM authors
ORDER BY name;

-- @name CreateAuthor :one
-- @param name TEXT NOT NULL
-- @param bio TEXT
INSERT INTO authors (
  name, bio
) VALUES (
  @name, @bio
)
RETURNING *;

-- @name GetAuthor
-- @param id BIGINT NOT NULL
DELETE FROM authors
WHERE id = @id;
```

## Future
The plan is to use similar annotations to support type annotations for output columns and values, Go type overrides for parameters and outputs, and JSON unmarshal / marshal hints.

**Comment by andrewmbenton** [comment]
In the named parameter example, I would expect the syntax to be
```
-- @name CreateAuthor :one
-- @param @name TEXT NOT NULL
-- @param @bio TEXT
```
so that it's consistent with the `$n` and `?n` treatment. Alternatively we could change the syntax in the numbered parameter cases to remove the `$` and `?` and just leave the numbers.

**Comment by kyleconroy** [comment]
I don't like the repetition of the `@` character, but agree that the `?` and `$` syntax feels unnecessary. We shouldn't allow named parameters to start with a number anyways.

**Comment by Emyrk** [comment]
Love this. Been type casting params for awhile now, but this is a much cleaner solution imo. :+1: 

Does this mean we no longer would need `sqlc.narg`?

**Comment by kyleconroy** [comment]
@Emyrk I just update the proposal with a slightly different syntax, which allows for specifying type information without change the nullability of a parameter.

**Comment by andrewmbenton** [comment]
> Does this mean we no longer would need `sqlc.narg`?

Yes, we just added a bit more clarity about that to the proposal.

**Comment by aliml92** [comment]
@kyleconroy @andrewmbenton Would it be okay to use generated model structs in place of query params? By having a syntax like this:
```
-- @name CreateProduct :one
-- @modelParam Product 
```
or any syntax that hints using model struct as params. I was able to achieve that by applyning new config option called `params_struct_overrides` in my forked repo. https://github.com/aliml92/sqlc/commit/d38d58a784b40e1c871e9083cf5ae3277a33ad60. Here are the sample codeK
1. `sqlc.yaml`:
```yaml
version: "2"
sql:
  - engine: "postgresql"
    queries: "./db/queries"
    schema: "./db/migrations"
    gen:
      go:
        package: "store"
        sql_package: "pgx/v5"
        out: "./internal/store"
        
        params_struct_overrides:
        - method_name: CreateProduct
          model_name: Product  
```
2. `models.go`
```go
type Product struct {
	ID         int64              `json:"id"`
	StoreID    int32              `json:"store_id"`
	CategoryID int32              `json:"category_id"`
	Name       string             `json:"name"`
	Brand      *string            `json:"brand"`
	Slug       string             `json:"slug"`
	ImageLinks []byte             `json:"image_links"`
	Specs      []byte             `json:"specs"`
	CreatedAt  pgtype.Timestamptz `json:"created_at"`
	UpdatedAt  pgtype.Timestamptz `json:"updated_at"`
}
```
3. `productcatalog.sql.go`
```go
 const createProduct = `-- name: CreateProduct :one
INSERT INTO products 
    (store_id, category_id, name, brand, slug, image_links, specs)
VALUES
    ($1, $2, $3, $4, $5, $6, $7)
RETURNING id
`

func (q *Queries) CreateProduct(ctx context.Context, p Product) (int64, error) {
	row := q.db.QueryRow(ctx, createProduct,
		p.StoreID,
		p.CategoryID,
		p.Name,
		p.Brand,
		p.Slug,
		p.ImageLinks,
		p.Specs,
	)
	var id int64
	err := row.Scan(&id)
	return id, err
}
 ```
However, enabling such override by type annotations would be much preferable though.

**Comment by Emyrk** [comment]
> However, enabling such override by type annotations would be much preferable though.

I also have wanted this feature, but there are potentially better ways to solve this (eg `sqlc.embed` as an example).

My fear of overrides is you lose some of the power of sqlc, which is making sure the models always match the query. If custom models are supported, is there anyway to add some "linting" or something that would warn the user when a new column is added and their custom model does not have it?

The query has all selected columns, so that would be possible. Just food for thought that custom structs might want some additional support to keep them "in line" with the sql.

**Comment by Emyrk** [comment]
:+1: again for this. Would be really helpful alongside type overrides for some edge case stuff. Currently trying to get tuples to work as parameters

**Comment by mmgfrcs** [comment]
Any progress on this? I was trying to annotate types for parameters in SQLite and found #3574, and this is a great solution to that issue.

**Comment by maddsua** [comment]
This would be particularly useful to let custom-defined types intended for the use together with json* functions to feel like first-class citizens in SQLITE.

Think of implementing an equivalent of Postgres'es `ANY()`. I feel like using `select value from json_each(your_custom_type_arg)` is a much cleaner solution than using `sqlc.slice`. Mainly because the slice doesn't always play well with other query arguments. Where's the only thing lacking for the json-based alternative is a proper type casting, so that the downstream code doesn't have to deal with any-typed query parameters.

**Comment by 1514599514** [comment]
```sql
SELECT CURRENT_TIMESTAMP::timestamptz AS current_tx_timestamp;
```
generate
```go
type Row struct {
	CurrentTxTimestamp pgtype.timestamptz
}
```

I couldn't find a way to reload the `CurrentTxTimestamp` field to the `time.Time` type.

It would be great if we could overload the return field type via annotations.
